libidset: replace internal nodeset class

garlick opened the pull request:
The old "nodeset" class has some deficiencies documented in #1495.  We created `libflux-idset` as an external, standalone library to export nodeset-like functionality to framework projects, but only created the bare minimum needed by flux-sched at the time and left nodeset in place.

In the spriit of recent API cleanup efforts, this PR finishes the job, adding functionality to idset, converting users, and removing nodeset.

garlick commented:
Still some squashing to do here before a review.  Sorry for the noise.

garlick commented:
Rebased.

One builder fail the valgrind test:

```
==963== 6,310,202 (224 direct, 6,309,978 indirect) bytes in 1 blocks are definitely lost in loss record 103 of 103
==963==    at 0x4C31D2F: realloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==963==    by 0x4E87BF9: ev_realloc (ev.c:1741)
==963==    by 0x4E881E4: ev_feed_event (ev.c:2005)
==963==    by 0x4E8B12F: queue_events (ev.c:2034)
==963==    by 0x4E8B12F: ev_run (ev.c:3714)
==963==    by 0x4E585C2: flux_reactor_run (reactor.c:140)
==963==    by 0xB7C910F: mod_main (job.c:938)
==963==    by 0x1142EB: module_thread (module.c:158)
==963==    by 0x55BA6DA: start_thread (pthread_create.c:463)
==963==    by 0x636188E: clone (clone.S:95)
==963== 
```

looks unrelated so I restarted it.

grondo commented:
Two quick comments, neither should be considered must-fix:

 * 73f7634: it would be slightly cleaner to add libidset to libflux_internal in a separate commit. Lumping it in with the mrpc update seems arbitrary and isn't symmetrical with the other code that is updated. (or maybe it is? I would also accept a comment in the commit message that describes why libidset was moved into libflux_internal in *this* commit)

 * e6b2e84 could use a commit message. The commit subject says the linkage was "fixed" but there isn't a hint why (which could be helpful for future developer wondering why that test cant use `$(test_ldadd)` like all the other tests)

garlick commented:
Thanks, good suggestions!

garlick commented:
OK, I think I addressed @grondo's comments.  I also expanded some commit messages a bit.

garlick commented:
Let me see if I can get the idset coverage up just a bit before this gets merged.

I also feel like the way everything is being linked is a little convoluted and might be improved if the veb stuff moved to libidset (though I've tried that twice now and got myself in the weeds both times).

garlick commented:
Rebased on current master and squashed down the unit test addition.  I think this could be ready to go in.  @chu11, sounds like @grondo may be offline for a bit.  Would you mind giving this a quick review?

chu11 commented on src/common/libidset/idset.h:
```diff
@@ -57,10 +65,44 @@ char *idset_encode (const struct idset *idset, int flags);
 int idset_set (struct idset *idset, unsigned int id);
+int idset_range_set (struct idset *idset, unsigned int lo, unsigned int hi);
+
+/* Remove id (or range [lo-hi] from idset.
```
stupid typo, no end paren ')'

chu11 commented on src/common/libidset/idset.c:
```diff
@@ -124,6 +157,124 @@ int idset_set (struct idset *idset, unsigned int id)
+
+    if (!idset1 || !idset2)
+        return false;
+
```
minor thought, would it be a net win to do some obvious checks before iterating through everything below?  I know ```vebsize()``` is costly and perhaps not worth it.  T.M != T.M?  I don't know the library that well, so maybe it's not a net win.

garlick replied:
`T.M` is the number of available slots in the set, not filled slots.  `vebsize()` is the size in bytes of the data structure, not related to filled slots.  So I don't think either of those is helpful here.  

In looking at this, I realized there are no users of `idset_compare()` so hey, let's just get rid of it.

chu11 commented on src/common/libidset/test/idset.c:
```diff
@@ -154,6 +413,12 @@ int main (int argc, char *argv[])
+    test_clear ();
+    test_range_clear ();
+    test_copy ();
 
```
i don't think i see a unit test for AUTOGROW?

garlick replied:
Thanks will add some.

garlick commented:
LMK how this looks and I'll squash the incremental devel.

chu11 commented on src/common/libidset/test/idset.c:
```diff
@@ -275,7 +275,7 @@ void test_range_set (void)
     ok (idset_range_set (idset, 0, 2) == 0,
         "idset_range_set 0-2 worked");
-    ok (idset_range_set (idset, 80, 79) == 0,
+    ok (idset_range_set (idset, 80, 79) == 0, // revsersed
```
typo "revsersed"

garlick replied:
Sheesh, I blame this keyboard.

garlick commented:
Fixed that typo and squashed.

chu11 commented:
lgtm, I'll hit the button once travis passes.  You'll have to rebase this or #1867, depending on what finishes first :-)

garlick commented:
Thanks!
